juju/framework.py

```diff
@@ -553,9 +570,13 @@ def __init__(self, parent, attr_name):
         except NoSnapshotError:
             data = StoredStateData(parent, attr_name)
 
+        # __dict__ is used to avoid infinite recursion.
         self.__dict__["_data"] = data
         self.__dict__["_attr_name"] = attr_name
 
+        # Has to be here because StoredStateData might be restored from persistent storage and it's __init__ will not be invoked.
```

Comment:
`__init__` will not be called in this case:

```
        try:
            data = parent.framework.load_snapshot(handle)
```

Whereas here it will be:

```
        except NoSnapshotError:
            data = StoredStateData(parent, attr_name)
```

The comment's purpose is to make it clear that this will not work as an alternative:

```
class StoredStateData(Object):

    on = StoredStateEvents()

    def __init__(self, parent, attr_name):
        super().__init__(parent, attr_name)
        self._cache = {}
        self.framework.observe(self.framework.on.pre_commit, self)  # <--- this
```